ESQL: Allow grouping key inside stats expressions

costin opened the pull request:
Similar to aggs, allow grouping keys to used inside STATS expressions by
 introducing a synthetic eval, e.g.:

STATS a = x + count(*) BY x becomes
STATS c = count(*) BY x | EVAL a = x + c | KEEP a, x

costin committed:
ESQL: Allow grouping key inside stats expressions

Similar to aggs, allow grouping keys to used inside STATS expressions by
 introducing a synthetic eval, e.g.:

STATS a = x + count(*) BY x becomes
STATS c = count(*) BY x | EVAL a = x + c | KEEP a, x

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

costin committed:
Update docs/changelog/106579.yaml

elasticsearchmachine commented:
Hi @costin, I've created a changelog YAML for you.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -1188,6 +1188,72 @@ e:i | l:i
+
+nestedAggsOverGroupingWithAliasAndProjection#[skip:-8.13.99,reason:supported in 8.14]
+FROM employees
+| STATS e = length(f) + 1, c = count(*) by f = first_name
```
On a similar note, I think we should also allow `| STATS e = f by f = first_name` if we allow `| STATS e = length(f) by f = first_name`; I think the current state of this PR does not enable this, right?

costin replied:
there's no value in STATS x by x; it will return x twice. just write STATS x.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -339,6 +347,60 @@ protected LogicalPlan doRule(LogicalPlan plan) {
+                        Attribute maybeResolved = maybeResolveAttribute(ua, resolvedList);
+                        if (maybeResolved != null) {
+                            changed.set(true);
+                            // ne = maybeResolved.resolved() ? resolved.get(maybeResolved) : maybeResolved;
```
Leftover?

alex-spies reviewed (approved):
LGTM - only minor remarks.

Cool stuff!

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -339,6 +347,60 @@ protected LogicalPlan doRule(LogicalPlan plan) {
+                for (Expression g : groupings) {
+                    Expression resolved = g.transformUp(UnresolvedAttribute.class, ua -> maybeResolveAttribute(ua, childrenOutput));
+                    if (resolved != g) {
+                        changed.set(true);
```
You don't need a `Holder` here, it's not used in lambdas. The `changed` is used further down, but here it's not actually needed.

costin replied:
the flag is used - since without setting (and then resetting), the grouping resolution gets lost.
When dealing with groups that provide aliases - first the groups have to be resolved, then the aggregates.
Instead of introducing a new bool, I've used the existing holder which is used in the lambda at 390.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -339,6 +347,60 @@ protected LogicalPlan doRule(LogicalPlan plan) {
+                groupings = newGroupings;
+                if (changed.get()) {
+                    a = new Aggregate(a.source(), a.child(), newGroupings, a.aggregates());
+                    changed.set(false);
```
This doesn't do anything.

costin replied:
it resets the flag so the aggregate resolution  (on line 398) gets the correct state.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -339,6 +347,60 @@ protected LogicalPlan doRule(LogicalPlan plan) {
+                    newAggregates.add(agg);
+                }
+
+                a = changed.get() ? new Aggregate(a.source(), a.child(), groupings, newAggregates) : a;
```
The code here (and in the method in general) is a bit hard to follow. Why the `new Aggregate(...)` is using `groupings` here? Why not `a.groupings()`? Earlier in the method there is this:

```
                groupings = newGroupings;
                if (changed.get()) {
                    a = new Aggregate(a.source(), a.child(), newGroupings, a.aggregates());
                    changed.set(false);
                }
```
It kinda gives me the feeling `a.groupings()` doesn't change and can safely be used here instead....

costin replied:
As you point out the groupings change and a new aggregate is created. but the groupings is kept around with the latest state - either it was changed or not.
groupings is used in 2 places through-out the same block, hence the intention to keep it consistent and use the same var instead of a.groupings() (see line 373 and 375).

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -157,6 +159,7 @@ private static void checkAggregate(LogicalPlan p, Set<Failure> failures, Attribu
         if (p instanceof Aggregate agg) {
 
```
Empty line

costin committed:
Update tests

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -1044,7 +1044,7 @@ FROM employees
 ;
 
 // tag::docsStatsByExpression-result[]
-my_count:long    |LEFT(last_name, 1):keyword
```
It aligns the header with the columns. I picked it up since the test was failing at some point.

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -1188,6 +1188,72 @@ e:i | l:i
 
+nestedAggsOverGroupingWithAlias#[skip:-8.13.99,reason:supported in 8.14]
+FROM employees
+| STATS e = length(f) + 1, count(*) by f = first_name
```
We should not allow the group key inside stats since its redundant, we already attach the key to the output.
Thanks for the queries - it looks like the Verifier could be improved.
I tried to keep things close to Postgres and, in case of expressions inside grouping, did not semantic matching rather basic equality.

`FROM employees | STATS  languages + emp_no BY languages + emp_no` works because the grouping key is easily found however we shouldn't allow this type of query.
`FROM employees | STATS e = languages + emp_no BY languages + emp_no` does not work because languages + emp_no is aliased and not matched with the grouping - better to do STATS e BY e = languages + emp_no.
However as I said, I'd prefer for us to prevent this and ideally have a better message ("the group key is already included in the output").

costin replied:
Added more tests and updated the behavior.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -339,6 +347,60 @@ protected LogicalPlan doRule(LogicalPlan plan) {
+                List<Expression> newGroupings = new ArrayList<>(groupings.size());
+                for (Expression g : groupings) {
+                    Expression resolved = g.transformUp(UnresolvedAttribute.class, ua -> maybeResolveAttribute(ua, childrenOutput));
+                    if (resolved != g) {
```
yes because either there's a transform or none. equality means checking the whole tree while we expect this to happen only once.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -339,6 +347,60 @@ protected LogicalPlan doRule(LogicalPlan plan) {
+                }
+            }
+
+            if (a.expressionsResolved() == false && Resolvables.resolved(groupings)) {
```
Not quite - the aggregates might be unresolved as well as the groupings. The condition is looking to investigate the aggregates only if the groupings are resolved.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -339,6 +347,60 @@ protected LogicalPlan doRule(LogicalPlan plan) {
+                        }
+                        return ne;
+                    });
+                    newAggregates.add(agg);
```
We do not. If we did, we would have to add it to the resolvedList so it becomes available to the next resolution call.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -165,7 +168,13 @@ private static void checkAggregate(LogicalPlan p, Set<Failure> failures, Attribu
                 });
-                nakedGroups.add(Alias.unwrap(e));
+                // remember naked expressions (to deal with implicit aliasing for grouping)
+                nakedGroups.add(e);
```
Updated the comment.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -176,27 +185,31 @@ private static void checkAggregate(LogicalPlan p, Set<Failure> failures, Attribu
+        List<Expression> groups,
+        AttributeSet groupRefs,
+        Set<Failure> failures,
+        int level
```
Will update the code to prevent the case of `STATS x by x`.

costin reviewed (commented):
added some clarifying comments.

costin committed:
Address feedback

alex-spies commented:
> added some clarifying comments.

I cannot see them as part of this PR - only updates to tests. Did you push that?

costin committed:
Address feedback

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java:
```diff
@@ -221,18 +225,33 @@ public PlanFactory visitStatsCommand(EsqlBaseParser.StatsCommandContext ctx) {
         }
         // grouping keys are automatically added as aggregations however the user is not allowed to specify them
         if (groupings.isEmpty() == false && aggregates.isEmpty() == false) {
-            var groupNames = new LinkedHashSet<>(Expressions.names(Expressions.references(groupings)));
+            var groupNames = new LinkedHashSet<>(Expressions.names(groupings));
+            var groupRefNames = new LinkedHashSet<>(Expressions.names(Expressions.references(groupings)));
 
             for (NamedExpression aggregate : aggregates) {
-                if (Alias.unwrap(aggregate) instanceof UnresolvedAttribute ua && groupNames.contains(ua.name())) {
-                    throw new ParsingException(ua.source(), "Cannot specify grouping expression [{}] as an aggregate", ua.name());
+                Expression e = Alias.unwrap(aggregate);
+                if (e.resolved() == false && e instanceof UnresolvedFunction == false) {
+                    String name = e.sourceText();
+                    if (groupNames.contains(name)) {
+                        fail(e, "Remove grouping key [{}] as is already included in the STATS output", name);
+                    } else if (groupRefNames.contains(name)) {
+                        fail(e, "Cannot specify grouping expression [{}] as an aggregate", name);
+                    }
                 }
             }
         }
-        aggregates.addAll(groupings);
+        // since groupings are aliased, add refs to it in the aggregates
+        for (Expression group : groupings) {
+            aggregates.add(Expressions.attribute(group));
+        }
+
         return input -> new Aggregate(source(ctx), input, new ArrayList<>(groupings), aggregates);
     }
 
+    private void fail(Expression exp, String message, Object... args) {
+        throw new VerificationException(Collections.singletonList(Failure.fail(exp, message, args)));
+    }
```
Improved the detection of grouping (or references) added to the aggregate. In case of failure a verification exception is thrown instead of a ParsingException for consistency reasons.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -91,7 +83,27 @@ public void testAggsExpressionsInStatsAggs() {
             "1:23: second argument of [count_distinct(languages, languages)] must be a constant, received [languages]",
             error("from test | stats x = count_distinct(languages, languages) by emp_no")
         );
+        // no agg function
+        assertEquals("1:19: expected an aggregate function but found [5]", error("from test | stats 5 by emp_no"));
+        // don't allow naked group
+        assertEquals(
+            "1:19: Remove grouping key [emp_no] as is already included in the STATS output",
+            error("from test | stats emp_no BY emp_no")
+        );
+        // don't allow naked group - even when it's an expression
+        assertEquals(
+            "1:19: Remove grouping key [languages + emp_no] as is already included in the STATS output",
+            error("from test | stats languages + emp_no BY languages + emp_no")
+        );
+        // don't allow group alias
+        assertEquals(
+            "1:19: Remove grouping key [e] as is already included in the STATS output",
+            error("from test | stats e BY e = languages + emp_no")
+        );
 
+        var message = error("from test | stats languages + emp_no BY e = languages + emp_no");
+        assertThat(message, containsString("column [languages] must appear in the STATS BY clause or be used in an aggregate function"));
+        assertThat(message, containsString("column [emp_no] must appear in the STATS BY clause or be used in an aggregate function"));
```
Tests checking the group cannot be used by itself.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -157,35 +157,36 @@ else if (p.resolved()) {
                 });
-                // remember naked expressions (to deal with implicit aliasing for grouping)
-                nakedGroups.add(e);
+                groups.add(e);
```
I still think that `groups` and `groupings` are identical because we add each element of `groupings` to `group` here.

I asserted `groups.equals(groupings)` here and it at least doesn't fail any verifier or csv test.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java:
```diff
@@ -221,11 +225,18 @@ public PlanFactory visitStatsCommand(EsqlBaseParser.StatsCommandContext ctx) {
+                if (e.resolved() == false && e instanceof UnresolvedFunction == false) {
+                    String name = e.sourceText();
+                    if (groupNames.contains(name)) {
+                        fail(e, "Remove grouping key [{}] as is already included in the STATS output", name);
```
```suggestion
                        fail(e, "Remove grouping key [{}] as it is already included in the STATS output", name);
```
`checkInvalidNamedExpressionUsage` in `Verifier.java` has the same typo.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -207,9 +208,9 @@ private static void checkInvalidNamedExpressionUsage(
-            // }
+            if (level == 0) {
+                failures.add(fail(e, "remove grouping key [{}] as is already included in the STATS output", e.sourceText()));
+            }
```
Do we need this, as it seems to be enforced in the `LogicalPlanBuilder` already?

Commenting out these lines still makes all verifier and csv tests pass.

astefan reviewed (commented):
Apologies, I am testing this first and look at the code after.

This is a series of queries that show, imo, some inconsistencies:
- `FROM employees | STATS e = languages + emp_no BY f = languages + emp_no`

`Found 2 problems\nline 1:28: column [languages] must appear in the STATS BY clause or be used in an aggregate function\nline 1:40: column [emp_no] must appear in the STATS BY clause or be used in an aggregate function`

The message is contradictory to the reality: both columns appear in the `STATS BY`.
- `FROM employees | STATS languages + emp_no BY languages + emp_no`

The message is different even though I removed the aliases: `Found 1 problem\nline 1:24: Remove grouping key [languages + emp_no] as is already included in the STATS output`

- `FROM employees | STATS languages + emp_no` (I followed the previous error message and removed the `BY` statement altogether)

`Found 2 problems\nline 1:24: column [languages] must appear in the STATS BY clause or be used in an aggregate function\nline 1:36: column [emp_no] must appear in the STATS BY clause or be used in an aggregate function`

This error message comes as a consequence of the user following the previous advice in the error message. And the one the user gets now gets him/her in the same situation as the previous query, like a loop.

- `FROM employees | STATS f = languages + emp_no BY f = languages + emp_no`

This query finally succeeds. But all the previous queries and error messages are really confusing; and notice the very small difference from the first query: the alias is the same here while in the first query the aliases were different.

costin committed:
Introduce EsqlAggregate

To better handle overriding aliases, introduce EsqlAggregate which keeps
 the declared structure intact during analysis and verification while
 merging the output. The deduplication happens now in the optimization
 phase.

costin committed:
Merge remote-tracking branch 'remotes/upstream/main' into esql/expressions-on-groups

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -663,59 +718,30 @@ private static class ResolveFunctions extends ParameterizedAnalyzerRule<LogicalP
-     * The duplication needs to be addressed to avoid ambiguity errors from commands further down
-     * the line.
-     */
-    private static class RemoveDuplicateProjections extends BaseAnalyzerRule {
```
Moved the rule into the optimizer.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/EsqlAggregate.java:
```diff
@@ -0,0 +1,60 @@
+ * a proper output.
+ * To simplify things, the Aggregate class will be replaced with a vanilla one.
+ */
+public class EsqlAggregate extends Aggregate {
```
Main change in order to preserve the changes into the Verifier and the move the duplicates away in the optimizer.
Might be something we'd want to do across the board for all elements that allow duplicates.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:
```diff
@@ -1177,9 +1179,14 @@ public void testAggsWithDuplicates() throws Exception {
         var order = as(limit.child(), OrderBy.class);
         var agg = as(order.child(), Aggregate.class);
         var aggregates = agg.aggregates();
-        assertThat(aggregates, hasSize(2));
-        assertThat(Expressions.names(aggregates), contains("x", "b"));
+        var output = agg.output();
+        assertThat(output, hasSize(2));
+        assertThat(Expressions.names(output), contains("x", "b"));
         var alias = as(aggregates.get(0), Alias.class);
+        var count = as(alias.child(), Count.class);
+        alias = as(aggregates.get(1), Alias.class);
+        var min = as(alias.child(), Min.class);
+        alias = as(aggregates.get(2), Alias.class);
```
Updated the tests to take into account the output instead of the aggs.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -91,7 +83,34 @@ public void testAggsExpressionsInStatsAggs() {
             "1:23: second argument of [count_distinct(languages, languages)] must be a constant, received [languages]",
             error("from test | stats x = count_distinct(languages, languages) by emp_no")
         );
+        // no agg function
+        assertEquals("1:19: expected an aggregate function but found [5]", error("from test | stats 5 by emp_no"));
+        // don't allow naked group
+        assertEquals("1:19: grouping key [emp_no] already specified in the STATS BY clause", error("from test | stats emp_no BY emp_no"));
+        // don't allow naked group - even when it's an expression
+        assertEquals(
+            "1:19: grouping key [languages + emp_no] already specified in the STATS BY clause",
+            error("from test | stats languages + emp_no BY languages + emp_no")
+        );
+        // don't allow group alias
+        assertEquals(
+            "1:19: grouping key [e] already specified in the STATS BY clause",
+            error("from test | stats e BY e = languages + emp_no")
+        );
 
+        var message = error("from test | stats languages + emp_no BY e = languages + emp_no");
+        assertThat(
+            message,
+            containsString(
+                "column [emp_no] cannot be used as an aggregate, once declared in the STATS BY grouping key [e = languages + emp_no]"
+            )
+        );
+        assertThat(
+            message,
+            containsString(
+                " column [languages] cannot be used as an aggregate, once declared in the STATS BY grouping key [e = languages + emp_no]"
+            )
```
@astefan incorporated the changes.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -293,6 +293,50 @@ public void testCombineProjectionWithAggregation() {
+     * \_Aggregate[[languages{f}#16],[MAX(emp_no{f}#13) AS x, languages{f}#16]]
+     *   \_EsRelation[test][_meta_field{f}#19, emp_no{f}#13, first_name{f}#14, ..]
+     */
+    public void testRemoveOverridesInAggregate() throws Exception {
+        var plan = plan("""
+            from test
+            | stats x = count(emp_no), x = min(emp_no), x = max(emp_no) by languages
+            | sort x
+            """);
+
+        var topN = as(plan, TopN.class);
+        var agg = as(topN.child(), Aggregate.class);
+        var aggregates = agg.aggregates();
+        assertThat(aggregates, hasSize(2));
+        assertThat(Expressions.names(aggregates), contains("x", "languages"));
+        var alias = as(aggregates.get(0), Alias.class);
+        var max = as(alias.child(), Max.class);
+    }
+
+    // expected stats b by b (grouping overrides the rest of the aggs)
+
+    /**
+     * Expects
+     * TopN[[Order[b{r}#10,ASC,LAST]],1000[INTEGER]]
+     * \_Aggregate[[b{r}#10],[languages{f}#16 AS b]]
+     *   \_EsRelation[test][_meta_field{f}#19, emp_no{f}#13, first_name{f}#14, ..]
+     */
+    public void testAggsWithOverridingInputAndGrouping() throws Exception {
+        var plan = plan("""
+            from test
+            | stats b = count(emp_no), b = max(emp_no) by b = languages
+            | sort b
+            """);
+
+        var topN = as(plan, TopN.class);
+        var agg = as(topN.child(), Aggregate.class);
+        var aggregates = agg.aggregates();
+        assertThat(aggregates, hasSize(1));
+        assertThat(Expressions.names(aggregates), contains("b"));
```
Preserved the initial tests from the analyzer here,  as they check the removed overrides.

alex-spies replied:
I think we should assert more than just the name here; this test would still pass if the `aggregates` consisted of `b = count(emp_no)` instead. We could check that `aggregates.get(0)` is an alias whose child has name `languages`.

astefan reviewed (commented):
This is more solid now. Thank you for rewriting some of the code.
I've tested this again and there are some aspects that seem inconsistent:
1. `FROM employees | STATS languages + emp_no BY languages + emp_no`

generates the following error message: `Found 1 problem\nline 1:24: grouping key [languages + emp_no] already specified in the STATS BY clause`

The following query (notice the different aliasing for the same expression) `FROM employees | STATS x = languages + emp_no BY y = languages + emp_no`

generates a different error message: `Found 2 problems\nline 1:28: column [languages] cannot be used as an aggregate, once declared in the STATS BY grouping key [y = languages + emp_no]\nline 1:40: column [emp_no] cannot be used as an aggregate, once declared in the STATS BY grouping key [y = languages + emp_no]`

Imo, the first error message should pretty much say the same thing as the second one, because is more clear: `cannot be used as an aggregate, once declared in the STATS BY grouping key`.

2. `FROM employees | STATS min(f) BY f = languages + emp_no` works fine (this is an aggregate function over an expression).

but `FROM employees | STATS min(languages + emp_no) BY languages + emp_no` doesn't work with the error message `Cannot work with target field [languages + emp_no] for agg [min(languages + emp_no)]`.

`FROM employees | eval x = languages + emp_no | STATS min(x) BY languages + emp_no` works again.

alex-spies reviewed (commented):
Since this PR allows duplicate names in the groups, I think we should add more csv tests as well to cover that - or tackle this in a follow-up.

I think there's a bug in `RemoveAggregateOverrides`;
The following query trips the logical verifier:
```
FROM employees
| STATS vals = COUNT(*) by x = emp_no, x = languages
```
To repro, just add as csv test.

IMHO, I'd prefer if we put `RemoveAggregateOverrides` into a separate PR, because overriding group names has multiple semantics that can be seen as correct. See my comments below.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -105,14 +110,7 @@ public class Analyzer extends ParameterizedRuleExecutor<LogicalPlan, AnalyzerCon
-            new ResolveFunctions(),
-            new RemoveDuplicateProjections()
-        );
+        var resolution = new Batch<>("Resolution", new ResolveTable(), new ResolveEnrich(), new ResolveFunctions(), new ResolveRefs());
```
`FunctionResolution` doesn't depend on the children, only on the function name. 
It should be possible to do this resolution once (instead of each call) and potentially move it at parsing time for very early validation.

costin committed:
Address feedback

costin committed:
Merge branch 'main' into esql/expressions-on-groups

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -1495,6 +1498,62 @@ private LogicalPlan rule(Eval eval) {
+     * The overrides appear when the same alias is used multiple times in aggregations and/or groupings:
+     * STATS x = COUNT(*), x = MIN(a) BY x = b + 1, x = c + 10
+     * becomes
+     * STATS BY x = c + 10
```
Thanks for pointing this out. The rule for removing the duplicate grouping was trying to be smart and thus failed and its job :) This has now been fixed.

The last occurrence overrides previous declarations. So in your example `STATS BY x = b + 1,  x = c + 10` becomes `STATS x = c + 10`.

costin commented:
@astefan I've updated the message to include your feedback for 1.
Re: 2 I couldn't reproduce the issue - I've added a [couple of tests](https://github.com/elastic/elasticsearch/pull/106579/files#diff-4a6f4603cd7b2634516d4db329da775479a30e6aae1adf3b049a79c25dfaa0a1R1215).

@alex-spies the grouping and agg overriding was there before this commit - see the tests moved into the LogicalPlanOptimizerTests from the AnalyzerTests (since the rule that changes EsqlAggregate to Aggregate moved from the Analyzer to the Optimizer).

costin committed:
Fix import

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -1188,6 +1188,19 @@ e:i | l:i
 4   | 3
 ;
 
+nestedAggsOverGroupingTwiceWithAlias#[skip:-8.12.99,reason:supported in 8.13]
```
Shouldn't this be
```suggestion
nestedAggsOverGroupingTwiceWithAlias#[skip:-8.13.99,reason:supported in 8.14]
```
?

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -1521,36 +1521,28 @@ protected LogicalPlan rule(Aggregate agg) {
-                else if (nameSet.add(aggName) == false) {
-                    newAggregates.remove(i);
+
+            groupings = removeDuplicateNames(groupings);
+            aggregates = removeDuplicateNames(aggregates);
```
Much easier to read this way, nice!

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:
```diff
@@ -1205,6 +1205,50 @@ public void testAggsWithOverridingInputAndGrouping() throws Exception {
+        var output = agg.output();
+        assertThat(output, hasSize(2));
+        var aggs = agg.aggregates();
+        var min = as(Alias.unwrap(aggs.get(0)), Min.class);
+        assertThat(min.arguments(), hasSize(1));
+        var group = Alias.unwrap(agg.groupings().get(0));
+        assertEquals(min.arguments().get(0), group);
```
nit: could assert that the group is a `Sum` of `emp_no` and `langauges`. Similarly in `testAggsOverGroupingKeyWithAlias` below.

alex-spies reviewed (commented):
> @alex-spies the grouping and agg overriding was there before this commit - see the tests moved into the LogicalPlanOptimizerTests from the AnalyzerTests (since the rule that changes EsqlAggregate to Aggregate moved from the Analyzer to the Optimizer).

Ok - I wasn't aware of this. In this case, I think this is good to go (after addressing Andrei's remarks).

costin committed:
Fix small bug that caused replacement of expressions inside aggregations
 to be skipped despite being applied
Improved Verifier to not repeat error messages in case for Aggregates
Removed verification heuristics for missing columns as functions as it
 was too broad

costin committed:
Merge remote-tracking branch 'remotes/upstream/main' into esql/expressions-on-groups

costin committed:
Fix tests

costin committed:
Fix for bwc test

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -1188,6 +1188,145 @@ e:i | l:i
 4   | 3
 ;
 
+nestedAggsOverGroupingExpressionWithoutAlias#[skip:-8.13.99,reason:supported in 8.14]
```
Added tests also as CSV spec

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -664,59 +719,30 @@ private static class ResolveFunctions extends ParameterizedAnalyzerRule<LogicalP
         @Override
         protected LogicalPlan rule(LogicalPlan plan, AnalyzerContext context) {
-            return plan.transformExpressionsUp(
+            return plan.transformExpressionsOnly(
```
Small optimization

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -87,16 +89,8 @@ else if (p.resolved()) {
                 p.forEachExpressionUp(Alias.class, a -> aliases.put(a.toAttribute(), a.child()));
                 return;
             }
-            // handle aggregate first to disambiguate between missing fields or incorrect function declaration
-            if (p instanceof Aggregate aggregate) {
-                for (NamedExpression agg : aggregate.aggregates()) {
-                    var child = Alias.unwrap(agg);
-                    if (child instanceof UnresolvedAttribute) {
-                        failures.add(fail(child, "invalid stats declaration; [{}] is not an aggregate function", child.sourceText()));
-                    }
-                }
-            }
-            p.forEachExpression(e -> {
```
This was a poor heuristic for cases where a function name is used instead of a field - however it back fired when dealing with missing fields so I've removed it. If we want it back we have to check it against the available functions by:
1. either marking the field inside the resolver (preferred)
2. doing this in the verifier.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -118,7 +112,20 @@ else if (p.resolved()) {
-            });
+            };
+
+            // aggregates duplicate grouping inside aggs - to avoid potentially confusing messages, we only check the aggregates
+            if (p instanceof Aggregate agg) {
+                // do groupings first
+                var groupings = agg.groupings();
+                groupings.forEach(unresolvedExpressions);
+                // followed by just the aggregates (to avoid going through the groups again)
+                var aggs = agg.aggregates();
+                int size = aggs.size() - groupings.size();
+                aggs.subList(0, size).forEach(unresolvedExpressions);
+            } else {
+                p.forEachExpression(unresolvedExpressions);
+            }
```
In cases where the grouping fails, the error message appears twice since the groups are added inside stats - the block above takes care of this.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
Need to revert the incorrect formatting in this file.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -3002,6 +3047,109 @@ public void testNestedMultiExpressionsInGroupingAndAggs() {
+     * \_Eval[[emp_no{f}#10 % 2[INTEGER] AS g, languages{f}#13 + emp_no{f}#10 % 2[INTEGER] AS $$emp_no_%_2_+_la>$COUNT$0]]
+     * \_EsRelation[test][_meta_field{f}#16, emp_no{f}#10, first_name{f}#11, ..]
+     */
+    public void testNestedExpressionsWithGroupingKeyInAggs() {
```
Added unit tests for the plans that were failing.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -1277,9 +1280,9 @@ protected LogicalPlan rule(Aggregate aggregate) {
                             Attribute attr = expToAttribute.computeIfAbsent(field.canonical(), k -> {
                                 Alias newAlias = new Alias(k.source(), syntheticName(k, af, counter[0]++), null, k, null, true);
                                 evals.add(newAlias);
-                                aggsChanged.set(true);
                                 return newAlias.toAttribute();
                             });
+                            aggsChanged.set(true);
```
This caused the bug @astefan discovered - when the replacement was done for an expression twice (both inside the grouping AND the aggregates), the flag was not set (it was only triggered for expressions that were not seen before).
This caused the old aggregations to be used and thus the change to be ignored.
This issue existed before this PR which exposed it.

costin reviewed (commented):
Left a number of comments.

costin reviewed (commented):
Left a number of comments.

costin committed:
Fix formatting

costin committed:
Attempt to make tests deterministic
